src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointMessageWriter.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Common;
+using Microsoft.Azure.SignalR.Protocol;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.SignalR
+{
+    /// <summary>
+    /// A service connection container which sends message to multiple service endpoints.
+    /// </summary>
+    internal class MultiEndpointMessageWriter : IMultiEndpointServiceConnectionContainer
+    {
+        private readonly ILogger _logger;
+        internal IReadOnlyCollection<HubServiceEndpoint> TargetEndpoints { get; }
+
+        public MultiEndpointMessageWriter(IEnumerable<HubServiceEndpoint> targetEndpoints, ILoggerFactory loggerFactory)
```

Comment:
parameter to use `IReadOnlyCollection` instead?

Reply:
I think `IEnumerable` as parameter type would reduce the caller effort. No matter what type the parameter is, the ctor takes the responsibility to convert it to read only collection.

Reply:
So `IReadOnlyCollection` as the parameter indicates the caller that the endpoints are immutable. And the caller decides when to evaluate it to an immutable list.